File: zoobin.js
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

let input = [];
rl.on('line', line => input.push(line));
rl.on('close', () => {
  const E = parseInt(input[0]);
  const normalize = (a, b) => a < b ? `${a}-${b}` : `${b}-${a}`;

  const initial = new Set();
  const expected = new Set();

  for (let i = 1; i <= E; i++) {
    const [a, b] = input[i].split(' ').map(Number);
    initial.add(normalize(a, b));
  }

  for (let i = E + 1; i <= 2 * E; i++) {
    const [a, b] = input[i].split(' ').map(Number);
    expected.add(normalize(a, b));
  }

  // Find edges that differ
  const diff = new Set();
  for (let edge of initial) {
    if (!expected.has(edge)) diff.add(edge);
  }
  for (let edge of expected) {
    if (!initial.has(edge)) diff.add(edge);
  }

  // Build graph from diff edges
  const graph = {};
  for (let edge of diff) {
    const [u, v] = edge.split('-').map(Number);
    if (!graph[u]) graph[u] = [];
    if (!graph[v]) graph[v] = [];
    graph[u].push(v);
    graph[v].push(u);
  }

  // Count cycles
  const visited = new Set();
  let cycles = 0;

  for (let node in graph) {
    node = parseInt(node);
    if (visited.has(node)) continue;

    const queue = [node];
    visited.add(node);
    let nodes = new Set([node]);
    let edgeCount = 0;

    while (queue.length > 0) {
      const u = queue.pop();
      for (let v of graph[u]) {
        edgeCount++;
        if (!visited.has(v)) {
          visited.add(v);
          queue.push(v);
          nodes.add(v);
        }
      }
    }

    if (edgeCount / 2 === nodes.size) cycles++;
  }

  console.log(cycles);
});

File: zoobin.py
from collections import defaultdict, deque

def normalize(a, b):
    return (min(a, b), max(a, b))

def read_edges(e, lines):
    edges = set()
    for line in lines:
        a, b = map(int, line.split())
        edges.add(normalize(a, b))
    return edges

def build_graph(edge_diff):
    graph = defaultdict(list)
    for u, v in edge_diff:
        graph[u].append(v)
        graph[v].append(u)
    return graph

def count_cycles(graph):
    visited = set()
    cycles = 0

    for node in graph:
        if node in visited:
            continue
        queue = deque([node])
        visited.add(node)
        nodes = set([node])
        edge_count = 0

        while queue:
            u = queue.popleft()
            for v in graph[u]:
                edge_count += 1
                if v not in visited:
                    visited.add(v)
                    queue.append(v)
                    nodes.add(v)

        if edge_count // 2 == len(nodes):  # simple cycle
            cycles += 1
    return cycles

# Input
E = int(input())
initial_lines = [input() for _ in range(E)]
expected_lines = [input() for _ in range(E)]

initial = read_edges(E, initial_lines)
expected = read_edges(E, expected_lines)

# Symmetric difference: edges that changed
diff = initial.symmetric_difference(expected)

# Build graph from changed edges
graph = build_graph(diff)

# Output: number of rotation cycles
print(count_cycles(graph))
